src/redis_disk.cc

```diff
@@ -0,0 +1,160 @@
+        break;
+      case RedisType::kRedisZSet:
+        s = GetZsetSize(ns_key, key_size);
+        break;
+      case RedisType::kRedisNone:
+        return rocksdb::Status::NotFound("Not found ", user_key);
+      case RedisType::kRedisStream:
+        return rocksdb::Status::NotSupported("Not support stream");
+    }
+    return s;
+}
+rocksdb::Status Disk::GetStringSize(const Slice &ns_key, uint64_t *key_size) {
+  auto key_range = rocksdb::Range(Slice(ns_key), Slice(ns_key.ToString() + static_cast<char>(0)));
+  return db_->GetApproximateSizes(option_, metadata_cf_handle_, &key_range, 1, key_size);
+}
+
+rocksdb::Status Disk::GetHashSize(const Slice &ns_key, uint64_t *key_size) {
+  HashMetadata metadata(false);
+  rocksdb::Status s = Database::GetMetadata(kRedisHash, ns_key, &metadata);
+  if (!s.ok()) return s.IsNotFound() ? rocksdb::Status::OK() : s;
+  return GetApproximateSizes(metadata, ns_key,
+                             storage_->GetCFHandle(Engine::kSubkeyColumnFamilyName),
+                             key_size);
+}
+
+
+rocksdb::Status Disk::GetSetSize(const Slice &ns_key, uint64_t *key_size) {
+  SetMetadata metadata(false);
+  rocksdb::Status s = Database::GetMetadata(kRedisSet, ns_key, &metadata);
+  if (!s.ok()) return s.IsNotFound() ? rocksdb::Status::OK() : s;
+  return GetApproximateSizes(metadata, ns_key,
+                             storage_->GetCFHandle(Engine::kSubkeyColumnFamilyName),
+                             key_size);
+}
+
+rocksdb::Status Disk::GetListSize(const Slice &ns_key, uint64_t *key_size) {
+  ListMetadata metadata(false);
+  rocksdb::Status s = Database::GetMetadata(kRedisList, ns_key, &metadata);
+  if (!s.ok()) return s.IsNotFound() ? rocksdb::Status::OK() : s;
+  std::string buf;
+  PutFixed64(&buf, metadata.head);
+  return GetApproximateSizes(metadata, ns_key,
+                             storage_->GetCFHandle(Engine::kSubkeyColumnFamilyName),
+                             key_size, buf);
+}
+
+rocksdb::Status Disk::GetZsetSize(const Slice &ns_key, uint64_t *key_size) {
+  ZSetMetadata metadata(false);
+  rocksdb::Status s = Database::GetMetadata(kRedisZSet, ns_key, &metadata);
+  if (!s.ok()) return s.IsNotFound() ? rocksdb::Status::OK() : s;
+  std::string score_bytes;
+  PutDouble(&score_bytes, kMinScore);
+  s = GetApproximateSizes(metadata, ns_key,
+                          storage_->GetCFHandle(Engine::kZSetScoreColumnFamilyName),
+                          key_size, score_bytes, score_bytes);
+  if (!s.ok()) return s;
+  return GetApproximateSizes(metadata, ns_key,
+                             storage_->GetCFHandle(Engine::kSubkeyColumnFamilyName),
+                             key_size);
+}
+
+rocksdb::Status Disk::GetBitmapSize(const Slice &ns_key, uint64_t *key_size) {
+  BitmapMetadata metadata(false);
+  rocksdb::Status s = Database::GetMetadata(kRedisBitmap, ns_key, &metadata);
+  if (!s.ok()) return s.IsNotFound() ? rocksdb::Status::OK() : s;
```

Comment:
As in https://github.com/apache/incubator-kvrocks/pull/882#discussion_r973585761,
you can put them out of the `switch`, then it can be eliminated from every method in this file.

Reply:
We need to set different types of metadata variables. The code seems to be more unreadable if the virtual function mechanism is used?

Reply:
Oh you are right, a factory function is needed for `Metadata`. So weird since the ctor of `Metadata` is public.